examples/weakref.cpp

```diff
@@ -0,0 +1,301 @@
+  if (!JS_IsExceptionPending(cx)) return;
+
+  JS::ExceptionStack exnStack{cx};
+  if (!JS::StealPendingExceptionStack(cx, &exnStack)) return;
+
+  JS::ErrorReportBuilder builder{cx};
+  if (!builder.init(cx, exnStack, JS::ErrorReportBuilder::NoSideEffects)) {
+    return;
+  }
+  JS::PrintError(stderr, builder, /* reportWarnings = */ false);
+}
+
+// This example integrates the FinalizationRegistry job queue together with the
+// Promise job handling, since that's a logical place that you might put it in
+// your embedding.
+//
+// However, it's not necessary to use JS::JobQueue and it's not necessary to
+// handle Promise jobs, in order to have FinalizationRegistry work. You do need
+// to have some kind of job queue, but it can be very minimal. It doesn't have
+// to be based on JS::JobQueue. The only requirement is that the enqueued
+// cleanup functions must be run "some time in the future".
+//
+// To approximate a minimal job queue, you might remove m_queue from this class
+// and remove the inheritance from JS::JobQueue and its overridden methods.
+class CustomJobQueue : public JS::JobQueue {
+ public:
+  explicit CustomJobQueue(JSContext* cx)
+      : m_queue(cx, js::SystemAllocPolicy{}),
+        m_finalizationRegistryCallbacks(cx),
+        m_draining(false) {}
+  ~CustomJobQueue() = default;
+
+  // JS::JobQueue override
+  JSObject* getIncumbentGlobal(JSContext* cx) override {
+    return JS::CurrentGlobalOrNull(cx);
+  }
+
+  // JS::JobQueue override
+  bool enqueuePromiseJob(JSContext* cx, JS::HandleObject promise,
+                         JS::HandleObject job, JS::HandleObject allocationSite,
+                         JS::HandleObject incumbentGlobal) override {
+    if (!m_queue.append(job)) {
+      JS_ReportOutOfMemory(cx);
+      return false;
+    }
+
+    JS::JobQueueMayNotBeEmpty(cx);
+    return true;
+  }
+
+  // JS::JobQueue override
+  void runJobs(JSContext* cx) override {
+    // Ignore nested calls of runJobs.
+    if (m_draining) {
+      return;
+    }
+
+    m_draining = true;
+
+    JS::Rooted<JSObject*> job{cx};
+    JS::Rooted<JS::Value> unused_rval{cx};
+
+    while (true) {
+      // Execute jobs in a loop until we've reached the end of the queue.
+      while (!m_queue.empty()) {
+        job = m_queue[0];
+        m_queue.erase(m_queue.begin());  // In production code, use a FIFO queue
+
+        // If the next job is the last job in the job queue, allow skipping the
+        // standard job queuing behavior.
+        if (m_queue.empty()) {
+          JS::JobQueueIsEmpty(cx);
+        }
+
+        JSAutoRealm ar{cx, job};
+        if (!JS::Call(cx, JS::UndefinedHandleValue, job,
+                      JS::HandleValueArray::empty(), &unused_rval)) {
+          // We can't throw the exception here, because there is nowhere to
+          // catch it. So, log it.
+          LogPendingException(cx);
+        }
+      }
+
+      // FinalizationRegistry callbacks may queue more microtasks, so only stop
```

Comment:
I wonder if this can say "job" instead of "microtask".
"microtask" is HTML-spec term, and other ECMAScript embedding doesn't have to use it,
and also given class and other part says "job", it would be better using the same term, here and below.